base_layer/wallet/src/util/encryption.rs

```diff
@@ -25,33 +25,86 @@ use aes_gcm::{
     Aes256Gcm,
 };
 use rand::{rngs::OsRng, RngCore};
+use tari_crypto::{
+    hash::blake2::Blake256,
+    hashing::{DomainSeparatedHasher, MacDomain},
+};
+use tari_utilities::ByteArray;
 
 pub const AES_NONCE_BYTES: usize = 12;
 pub const AES_KEY_BYTES: usize = 32;
+pub const AES_MAC_BYTES: usize = 32;
 
 pub trait Encryptable<C> {
+    fn source_key(&self, field_name: &'static str) -> Vec<u8>;
     fn encrypt(&mut self, cipher: &C) -> Result<(), String>;
     fn decrypt(&mut self, cipher: &C) -> Result<(), String>;
 }
 
-pub fn decrypt_bytes_integral_nonce(cipher: &Aes256Gcm, ciphertext: Vec<u8>) -> Result<Vec<u8>, String> {
-    if ciphertext.len() < AES_NONCE_BYTES {
+pub fn decrypt_bytes_integral_nonce(
+    cipher: &Aes256Gcm,
+    source_key: Vec<u8>,
+    ciphertext: Vec<u8>,
+) -> Result<Vec<u8>, String> {
+    if ciphertext.len() < AES_NONCE_BYTES + AES_MAC_BYTES {
         return Err(AeadError.to_string());
     }
-    let (nonce, cipher_text) = ciphertext.split_at(AES_NONCE_BYTES);
+
+    let (nonce, ciphertext) = ciphertext.split_at(AES_NONCE_BYTES);
+    let (ciphertext, stored_mac) = ciphertext.split_at(ciphertext.len() - AES_MAC_BYTES);
```

Comment:
replace with `check_sub` or `saturating_sub` otherwise this could possibly cause a panic